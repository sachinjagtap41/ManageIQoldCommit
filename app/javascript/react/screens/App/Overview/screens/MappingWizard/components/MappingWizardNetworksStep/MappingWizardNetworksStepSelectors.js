export const groupByUidEms = (networks = []) =>
  networks.reduce(
    (networksMap, network) => ({
      ...networksMap,
      [network.uid_ems]: networksMap[network.uid_ems]
        ? [...networksMap[network.uid_ems], network]
        : [network]
    }),
    {}
  );
