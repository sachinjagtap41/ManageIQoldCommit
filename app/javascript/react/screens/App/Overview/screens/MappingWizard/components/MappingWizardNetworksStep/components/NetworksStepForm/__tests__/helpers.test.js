import {
  getRepresentatives,
  sourceNetworksFilter,
  clustersMappingWithTreeViewAttrs,
  targetNetworkWithTreeViewAttrs,
  sourceNetworkWithTreeViewAttrs,
  networkGroupingForRep,
  dedupeMappedSourceNetworks,
  dedupeMappedTargetNetworks,
  mappingsForTreeView,
  mappingWithTargetNetworkRemoved,
  mappingWithSourceNetworkRemoved
} from '../helpers';
import { groupByUidEms } from '../../../MappingWizardNetworksStepSelectors';
import {
  sourceNetworks,
  targetNetworks,
  sourceCluster,
  targetCluster,
  clustersMapping,
  networkGrouping
} from '../networksStepForm.fixtures';

test('getRepresentatives should return an array with a representative from each source network grouping', () => {
  expect(getRepresentatives(groupByUidEms(sourceNetworks))).toMatchSnapshot();
});

test('sourceNetworksFilter should return a representative for each source network grouping that have not been mapped', () => {
  const networksStepMapping = {
    ...targetCluster,
    nodes: [
      {
        ...targetNetworks[0],
        nodes: [
          {
            ...sourceNetworks[2]
          }
        ]
      }
    ]
  };
  expect(
    sourceNetworksFilter(groupByUidEms(sourceNetworks), [networksStepMapping])
  ).toMatchSnapshot();
});

test('clustersMappingWithTreeViewAttrs adds extra attributes for display in TreeView', () => {
  expect(clustersMappingWithTreeViewAttrs(clustersMapping)).toMatchSnapshot();
});

test('targetNetworkWithTreeViewAttrs adds extra attributes for display in TreeView', () => {
  expect(targetNetworkWithTreeViewAttrs(targetNetworks[0])).toMatchSnapshot();
});

test('sourceNetworkWithTreeViewAttrs adds extra attributes for display in TreeView', () => {
  const [sourceNetwork] = sourceNetworks;
  const selectedCluster = sourceCluster;

  expect(
    sourceNetworkWithTreeViewAttrs(sourceNetwork, selectedCluster)
  ).toMatchSnapshot();
});

test('networkGroupingForRep returns the network grouping, with TreeView attrs, for a representative', () => {
  const [sourceNetworkRep] = networkGrouping;
  const groupedSourceNetworks = groupByUidEms(networkGrouping);
  const selectedCluster = sourceCluster;

  expect(
    networkGroupingForRep(
      sourceNetworkRep,
      groupedSourceNetworks,
      selectedCluster
    )
  ).toMatchSnapshot();
});

test("dedupeMappedSourceNetworks reduces a network mapping's source networks to representatives", () => {
  const networksMapping = {
    ...targetNetworks[0],
    nodes: networkGrouping
  };

  expect(dedupeMappedSourceNetworks(networksMapping)).toMatchSnapshot();
});

test("dedupeMappedTargetNetworks reduces a networks step mapping's nodes to representatives of each target network grouping", () => {
  const [targetNetworkSwitchOne, targetNetworkSwitchTwo] = targetNetworks;
  const networksStepMapping = {
    ...targetCluster,
    nodes: [targetNetworkSwitchOne, targetNetworkSwitchTwo]
  };

  expect(dedupeMappedTargetNetworks(networksStepMapping).nodes).toHaveLength(1);
});

test('mappingsForTreeView reduces source networks to representatives for all networks step mappings', () => {
  const networksStepMapping = {
    ...targetCluster,
    nodes: [
      {
        ...targetNetworks[0],
        nodes: networkGrouping
      }
    ]
  };

  expect(mappingsForTreeView([networksStepMapping])).toMatchSnapshot();
});

describe('mappingWithTargetNetworkRemoved', () => {
  const [targetNetworkToRemove, targetNetworkShouldRemain] = targetNetworks;
  const [sourceNetworkOne, sourceNetworkTwo] = sourceNetworks;
  const nodeToRemove = {
    ...targetNetworkToRemove,
    nodes: [sourceNetworkOne]
  };

  test('removes the networks mapping for the specified target network', () => {
    const nodeShouldRemain = {
      ...targetNetworkShouldRemain,
      nodes: [sourceNetworkTwo]
    };
    const networksStepMapping = {
      ...targetCluster,
      nodes: [nodeToRemove, nodeShouldRemain]
    };

    expect(
      mappingWithTargetNetworkRemoved(
        networksStepMapping,
        targetNetworkToRemove
      )
    ).toMatchSnapshot();
  });

  test('returns null if no networks mappings remain', () => {
    const networksStepMapping = {
      ...targetCluster,
      nodes: [nodeToRemove]
    };

    expect(
      mappingWithTargetNetworkRemoved(
        networksStepMapping,
        targetNetworkToRemove
      )
    ).toMatchSnapshot();
  });
});

describe('mappingWithSourceNetworkRemoved', () => {
  const [, , sourceNetworkShouldRemain] = sourceNetworks;
  const [sourceNetworkToRemove] = networkGrouping;
  const [targetNetwork] = targetNetworks;

  test('removes all networks whose uid_ems matches that of the specified source network', () => {
    const networksMapping = {
      ...targetNetwork,
      nodes: [...networkGrouping, sourceNetworkShouldRemain]
    };

    expect(
      mappingWithSourceNetworkRemoved(networksMapping, sourceNetworkToRemove)
    ).toMatchSnapshot();
  });

  test('returns null if no source networks remain', () => {
    const networksMapping = {
      ...targetNetwork,
      nodes: [sourceNetworkToRemove]
    };

    expect(
      mappingWithSourceNetworkRemoved(networksMapping, sourceNetworkToRemove)
    ).toMatchSnapshot();
  });
});
