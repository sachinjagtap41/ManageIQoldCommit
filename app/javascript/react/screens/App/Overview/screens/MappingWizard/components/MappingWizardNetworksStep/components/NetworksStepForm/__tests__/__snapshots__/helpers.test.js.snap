// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clustersMappingWithTreeViewAttrs adds extra attributes for display in TreeView 1`] = `
Object {
  "created_on": "2017-10-06T16:41:09Z",
  "drs_automation_level": "fullyAutomated",
  "drs_enabled": false,
  "drs_migration_threshold": 3,
  "effective_cpu": 109356,
  "effective_memory": 790163881984,
  "ems_id": "1",
  "ems_ref": "domain-c7",
  "ems_ref_obj": "domain-c7",
  "ha_admit_control": true,
  "ha_enabled": false,
  "ha_max_failures": 1,
  "href": "http://localhost:3000/api/clusters/1",
  "id": "1",
  "last_perf_capture_on": null,
  "name": "RHV Cluster 1",
  "nodes": Array [
    Object {
      "created_on": "2017-10-06T16:41:09Z",
      "drs_automation_level": "fullyAutomated",
      "drs_enabled": false,
      "drs_migration_threshold": 3,
      "effective_cpu": 109356,
      "effective_memory": 790163881984,
      "ems_id": "1",
      "ems_ref": "domain-c7",
      "ems_ref_obj": "domain-c7",
      "ha_admit_control": true,
      "ha_enabled": false,
      "ha_max_failures": 1,
      "href": "http://localhost:3000/api/clusters/1",
      "icon": "fa fa-file-o",
      "id": "1",
      "last_perf_capture_on": null,
      "name": "VMWare Cluster 1",
      "text": "VMWare Cluster 1",
      "type": null,
      "uid_ems": "domain-c7",
      "updated_on": "2018-01-20T23:12:00Z",
    },
  ],
  "selectable": false,
  "selected": false,
  "state": Object {
    "expanded": true,
  },
  "text": "RHV Cluster 1",
  "type": null,
  "uid_ems": "domain-c7",
  "updated_on": "2018-01-20T23:12:00Z",
}
`;

exports[`dedupeMappedSourceNetworks reduces a network mapping's source networks to representatives 1`] = `
Object {
  "allow_promiscuous": null,
  "computed_allow_promiscuous": null,
  "computed_forged_transmits": null,
  "computed_mac_changes": null,
  "created_on": "2017-06-01T04:24:18Z",
  "forged_transmits": null,
  "href": "http://localhost:3000/api/lans/10000000000019",
  "id": "1-1",
  "mac_changes": null,
  "name": "RHV Cluster 1 - ovirtmgmt1",
  "nodes": Array [
    Object {
      "allow_promiscuous": null,
      "computed_allow_promiscuous": null,
      "computed_forged_transmits": null,
      "computed_mac_changes": null,
      "created_on": "2017-06-01T04:24:18Z",
      "forged_transmits": null,
      "href": "http://localhost:3000/api/lans/10000000000007",
      "id": "1-1",
      "mac_changes": null,
      "name": "VMWareCluster1-vm_network1",
      "parent_id": null,
      "switch_id": "1",
      "tag": "23",
      "uid_ems": "VMWareCluster1-vm_network1",
      "updated_on": "2017-06-01T04:24:18Z",
    },
  ],
  "parent_id": null,
  "switch_id": "10000000000000",
  "tag": null,
  "uid_ems": "00000000-0000-0000-0000-000000000000",
  "updated_on": "2017-06-01T04:24:18Z",
}
`;

exports[`getRepresentatives should return an array with a representative from each source network grouping 1`] = `
Array [
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000007",
    "id": "1-1",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network1",
    "parent_id": null,
    "switch_id": "1",
    "tag": "23",
    "uid_ems": "VMWareCluster1-vm_network1",
    "updated_on": "2017-06-01T04:24:18Z",
  },
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000009",
    "id": "1-3",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network2",
    "parent_id": null,
    "switch_id": "1",
    "tag": "24",
    "uid_ems": "VMWareCluster1-vm_network2",
    "updated_on": "2017-06-01T04:24:18Z",
  },
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000010",
    "id": "1-4",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network3",
    "parent_id": null,
    "switch_id": "1",
    "tag": "24",
    "uid_ems": "VMWareCluster1-vm_network3",
    "updated_on": "2017-06-01T04:24:18Z",
  },
]
`;

exports[`mappingWithSourceNetworkRemoved removes all networks whose uid_ems matches that of the specified source network 1`] = `
Object {
  "allow_promiscuous": null,
  "computed_allow_promiscuous": null,
  "computed_forged_transmits": null,
  "computed_mac_changes": null,
  "created_on": "2017-06-01T04:24:18Z",
  "forged_transmits": null,
  "href": "http://localhost:3000/api/lans/10000000000019",
  "id": "1-1",
  "mac_changes": null,
  "name": "RHV Cluster 1 - ovirtmgmt1",
  "nodes": Array [
    Object {
      "allow_promiscuous": null,
      "computed_allow_promiscuous": null,
      "computed_forged_transmits": null,
      "computed_mac_changes": null,
      "created_on": "2017-06-01T04:24:18Z",
      "forged_transmits": null,
      "href": "http://localhost:3000/api/lans/10000000000009",
      "id": "1-3",
      "mac_changes": null,
      "name": "VMWareCluster1-vm_network2",
      "parent_id": null,
      "switch_id": "1",
      "tag": "24",
      "uid_ems": "VMWareCluster1-vm_network2",
      "updated_on": "2017-06-01T04:24:18Z",
    },
  ],
  "parent_id": null,
  "switch_id": "10000000000000",
  "tag": null,
  "uid_ems": "00000000-0000-0000-0000-000000000000",
  "updated_on": "2017-06-01T04:24:18Z",
}
`;

exports[`mappingWithSourceNetworkRemoved returns null if no source networks remain 1`] = `null`;

exports[`mappingWithTargetNetworkRemoved removes the networks mapping for the specified target network 1`] = `
Object {
  "created_on": "2017-10-06T16:41:09Z",
  "drs_automation_level": "fullyAutomated",
  "drs_enabled": false,
  "drs_migration_threshold": 3,
  "effective_cpu": 109356,
  "effective_memory": 790163881984,
  "ems_id": "1",
  "ems_ref": "domain-c7",
  "ems_ref_obj": "domain-c7",
  "ha_admit_control": true,
  "ha_enabled": false,
  "ha_max_failures": 1,
  "href": "http://localhost:3000/api/clusters/1",
  "id": "1",
  "last_perf_capture_on": null,
  "name": "RHV Cluster 1",
  "nodes": Array [
    Object {
      "allow_promiscuous": null,
      "computed_allow_promiscuous": null,
      "computed_forged_transmits": null,
      "computed_mac_changes": null,
      "created_on": "2017-06-01T04:24:19Z",
      "forged_transmits": null,
      "href": "http://localhost:3000/api/lans/10000000000020",
      "id": "1-2",
      "mac_changes": null,
      "name": "RHV Cluster 1 - ovirtmgmt1",
      "nodes": Array [
        Object {
          "allow_promiscuous": null,
          "computed_allow_promiscuous": null,
          "computed_forged_transmits": null,
          "computed_mac_changes": null,
          "created_on": "2017-06-01T04:24:18Z",
          "forged_transmits": null,
          "href": "http://localhost:3000/api/lans/10000000000008",
          "id": "1-2",
          "mac_changes": null,
          "name": "VMWareCluster1-vm_network1",
          "parent_id": null,
          "switch_id": "2",
          "tag": "24",
          "uid_ems": "VMWareCluster1-vm_network1",
          "updated_on": "2017-06-01T04:24:18Z",
        },
      ],
      "parent_id": null,
      "switch_id": "10000000000001",
      "tag": null,
      "uid_ems": "00000000-0000-0000-0000-000000000000",
      "updated_on": "2017-06-01T04:24:19Z",
    },
  ],
  "type": null,
  "uid_ems": "domain-c7",
  "updated_on": "2018-01-20T23:12:00Z",
}
`;

exports[`mappingWithTargetNetworkRemoved returns null if no networks mappings remain 1`] = `null`;

exports[`mappingsForTreeView reduces source networks to representatives for all networks step mappings 1`] = `
Array [
  Object {
    "created_on": "2017-10-06T16:41:09Z",
    "drs_automation_level": "fullyAutomated",
    "drs_enabled": false,
    "drs_migration_threshold": 3,
    "effective_cpu": 109356,
    "effective_memory": 790163881984,
    "ems_id": "1",
    "ems_ref": "domain-c7",
    "ems_ref_obj": "domain-c7",
    "ha_admit_control": true,
    "ha_enabled": false,
    "ha_max_failures": 1,
    "href": "http://localhost:3000/api/clusters/1",
    "id": "1",
    "last_perf_capture_on": null,
    "name": "RHV Cluster 1",
    "nodes": Array [
      Object {
        "allow_promiscuous": null,
        "computed_allow_promiscuous": null,
        "computed_forged_transmits": null,
        "computed_mac_changes": null,
        "created_on": "2017-06-01T04:24:18Z",
        "forged_transmits": null,
        "href": "http://localhost:3000/api/lans/10000000000019",
        "id": "1-1",
        "mac_changes": null,
        "name": "RHV Cluster 1 - ovirtmgmt1",
        "nodes": Array [
          Object {
            "allow_promiscuous": null,
            "computed_allow_promiscuous": null,
            "computed_forged_transmits": null,
            "computed_mac_changes": null,
            "created_on": "2017-06-01T04:24:18Z",
            "forged_transmits": null,
            "href": "http://localhost:3000/api/lans/10000000000007",
            "id": "1-1",
            "mac_changes": null,
            "name": "VMWareCluster1-vm_network1",
            "parent_id": null,
            "switch_id": "1",
            "tag": "23",
            "uid_ems": "VMWareCluster1-vm_network1",
            "updated_on": "2017-06-01T04:24:18Z",
          },
        ],
        "parent_id": null,
        "switch_id": "10000000000000",
        "tag": null,
        "uid_ems": "00000000-0000-0000-0000-000000000000",
        "updated_on": "2017-06-01T04:24:18Z",
      },
    ],
    "type": null,
    "uid_ems": "domain-c7",
    "updated_on": "2018-01-20T23:12:00Z",
  },
]
`;

exports[`networkGroupingForRep returns the network grouping, with TreeView attrs, for a representative 1`] = `
Array [
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000007",
    "icon": "fa fa-file-o",
    "id": "1-1",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network1",
    "parent_id": null,
    "selectable": true,
    "selected": false,
    "sourceClusterId": "1",
    "switch_id": "1",
    "tag": "23",
    "text": "VMWareCluster1-vm_network1",
    "uid_ems": "VMWareCluster1-vm_network1",
    "updated_on": "2017-06-01T04:24:18Z",
  },
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000008",
    "icon": "fa fa-file-o",
    "id": "1-2",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network1",
    "parent_id": null,
    "selectable": true,
    "selected": false,
    "sourceClusterId": "1",
    "switch_id": "2",
    "tag": "24",
    "text": "VMWareCluster1-vm_network1",
    "uid_ems": "VMWareCluster1-vm_network1",
    "updated_on": "2017-06-01T04:24:18Z",
  },
]
`;

exports[`sourceNetworkWithTreeViewAttrs adds extra attributes for display in TreeView 1`] = `
Object {
  "allow_promiscuous": null,
  "computed_allow_promiscuous": null,
  "computed_forged_transmits": null,
  "computed_mac_changes": null,
  "created_on": "2017-06-01T04:24:18Z",
  "forged_transmits": null,
  "href": "http://localhost:3000/api/lans/10000000000007",
  "icon": "fa fa-file-o",
  "id": "1-1",
  "mac_changes": null,
  "name": "VMWareCluster1-vm_network1",
  "parent_id": null,
  "selectable": true,
  "selected": false,
  "sourceClusterId": "1",
  "switch_id": "1",
  "tag": "23",
  "text": "VMWareCluster1-vm_network1",
  "uid_ems": "VMWareCluster1-vm_network1",
  "updated_on": "2017-06-01T04:24:18Z",
}
`;

exports[`sourceNetworksFilter should return a representative for each source network grouping that have not been mapped 1`] = `
Array [
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000007",
    "id": "1-1",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network1",
    "parent_id": null,
    "switch_id": "1",
    "tag": "23",
    "uid_ems": "VMWareCluster1-vm_network1",
    "updated_on": "2017-06-01T04:24:18Z",
  },
  Object {
    "allow_promiscuous": null,
    "computed_allow_promiscuous": null,
    "computed_forged_transmits": null,
    "computed_mac_changes": null,
    "created_on": "2017-06-01T04:24:18Z",
    "forged_transmits": null,
    "href": "http://localhost:3000/api/lans/10000000000010",
    "id": "1-4",
    "mac_changes": null,
    "name": "VMWareCluster1-vm_network3",
    "parent_id": null,
    "switch_id": "1",
    "tag": "24",
    "uid_ems": "VMWareCluster1-vm_network3",
    "updated_on": "2017-06-01T04:24:18Z",
  },
]
`;

exports[`targetNetworkWithTreeViewAttrs adds extra attributes for display in TreeView 1`] = `
Object {
  "allow_promiscuous": null,
  "computed_allow_promiscuous": null,
  "computed_forged_transmits": null,
  "computed_mac_changes": null,
  "created_on": "2017-06-01T04:24:18Z",
  "forged_transmits": null,
  "href": "http://localhost:3000/api/lans/10000000000019",
  "id": "1-1",
  "mac_changes": null,
  "name": "RHV Cluster 1 - ovirtmgmt1",
  "parent_id": null,
  "selectable": true,
  "selected": false,
  "state": Object {
    "expanded": true,
  },
  "switch_id": "10000000000000",
  "tag": null,
  "text": "RHV Cluster 1 - ovirtmgmt1",
  "uid_ems": "00000000-0000-0000-0000-000000000000",
  "updated_on": "2017-06-01T04:24:18Z",
}
`;
